docs/EVE-IMAGE-REPRODUCIBILITY.md

```diff
@@ -0,0 +1,274 @@
+
+Further, if the directory is git "dirty", i.e. has any uncommitted files that are not in `.gitignore`, or has changed but uncommitted files,
+then the image tag is appended with the word `dirty` and the hash of the contents of the directory, e.g.
+`b20146debd08bbfffb23f53fc9f5c3c4af377f67-dirty-abcdefg112334`.
+
+The sources to all components used in the image, based on the `Dockerfile`, must come from one of several places.
+
+|Source|Method|Reproducibility|
+|---|---|---|
+|Another container image|Using `FROM` in the Dockerfile|Must use a properly immutably tagged image or one that includes the content hash|
+|Internet|`curl`, `wget` or other download|Must download a specific version and, wherever possible, check the content hash, md5 or sha|
+|Go packages|`go mod` or `go get`|Go inherently records the specific hash of the package version used in `go.sum`, which itself is git committed|
+|OS packages|`apk add`|See next section|
+
+It is possible for human error in any of the above to slip through. EVE uses peer review and maintainer approval to reduce the probabilities of
+any non-reproducible source slipping in.
+
+Further, the EVE project is exploring additional gates. These include:
+
+* Contributors certifying on each Pull Request that they have not used non-reproducible sources.
+* Regular automated audits of all `pkg/` sources.
+* CI scans of `pkg/` sources.
+
+## Linux Packages
+
+The primary source for many software packages in Linux is the OS packaging system. All of the OCI images used in EVE are based on
+[Alpine](https://hub.docker.com/_/alpine), which uses the [apk package manager](https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper).
+
+For example, to install bash:
+
+```sh
+apk add bash
+```
+
+Alpine packages creates unique challenges for reproducibility. Specifically:
+
+1. The version of the package changes when changing the alpine version. `apk add bash` can give entirely different bits when running on `alpine:3.15` vs `alpine:3.16`.
+1. The version of the package can change even _without_ changing the alpine version. `apk add bash` can give a different version running today on `alpine:3.16` vs yesterday or tomorrow on the same base `alpine:3.16`.
+
+It is possible to mitigate, but not completely solve, the above issue by adding a specific version:
+
+```sh
+apk add bash=5.1.16-r2
+```
+
+The above does not completely solve it, as sometimes alpine packages change their bits _even for the same version_. Thus, running
+`apk add bash=5.1.16-r2` on the same base `alpine:3.16` can give different bits on different days. Alpine packages are changed "under the covers"
+when bugs are found or updates are provided.
+
+In order to resolve this issue, both the external/upstream `linuxkit/*` packages and the `lfedge/eve-*` packages in this repository follow the same principles.
+
+1. Create an alpine "base" image: `linuxkit/alpine` and `lfedge/eve-alpine`
```

Comment:
Given that we have lfedge/eve-alpine-base as well, should we call this the alpine "cache" image? We informally refer to this as a cache.

Reply:
We do have a bit of a naming overload issue, do we not?

I can rename most of these to "cache". My concern was that a cache implies something "on the side", that I choose to use, where these are `FROM`, i.e. something I build on top of. It isn't that my downstream package builds on `eve-alpine-base` and somewhere configures "use this other image as my cache". Rather, it is that downstream uses `eve-alpine` _as its base_ via `FROM lfedge/eve-alpine`.

What do you think?

Reply:
I updated it, I think this is clearer.